lib/eager_group/preloader.rb

```diff
@@ -22,37 +26,44 @@ def run
         end
-        record_ids = @records.map { |record| record.send(primary_key) }
-        unless definition = @klass.eager_group_definitions[definition_key]
-          next
+
+        Array.wrap(definition_key).each do |key|
+          find_aggregate_values_per_definition!(key, arguments)
         end
+      end
+    end
 
-        reflection = @klass.reflect_on_association(definition.association)
-        association_class = reflection.klass
-        association_class = association_class.instance_exec(*arguments, &definition.scope) if definition.scope
+    def find_aggregate_values_per_definition!(definition_key, arguments)
+      unless definition = @klass.eager_group_definitions[definition_key]
+        return
+      end
 
-        foreign_key, aggregate_hash =
-          if reflection.is_a?(ActiveRecord::Reflection::HasAndBelongsToManyReflection)
-            ["#{reflection.join_table}.#{reflection.foreign_key}", @klass.joins(reflection.name)]
-          elsif reflection.through_reflection
-            [
-              "#{reflection.through_reflection.name}.#{reflection.through_reflection.foreign_key}",
-              @klass.joins(reflection.name)
-            ]
-          else
-            [reflection.foreign_key, association_class]
-          end
-        aggregate_hash =
-          aggregate_hash.where(foreign_key => record_ids).where(polymophic_as_condition(reflection)).group(foreign_key)
-            .send(definition.aggregation_function, definition.column_name)
-        if definition.need_load_object
-          aggregate_objects = reflection.klass.find(aggregate_hash.values).each_with_object({}) { |o, h| h[o.id] = o }
-          aggregate_hash.keys.each { |key| aggregate_hash[key] = aggregate_objects[aggregate_hash[key]] }
-        end
-        @records.each do |record|
-          id = record.send(primary_key)
-          record.send("#{definition_key}=", aggregate_hash[id] || definition.default_value)
+      reflection = @klass.reflect_on_association(definition.association)
+      return if reflection.blank?
+
+      aggregation_finder_class = if reflection.is_a?(ActiveRecord::Reflection::HasAndBelongsToManyReflection)
+        ManyToMany
+      elsif reflection.through_reflection
+        if reflection.through_reflection.is_a?(ActiveRecord::Reflection::BelongsToReflection)
+          HasManyThroughBelongsTo
+        else
+          HasManyThroughMany
         end
+      else
+        HasMany
+      end
+
+      aggregation_finder = aggregation_finder_class.new(@klass, definition, arguments, @records)
+      aggregate_hash = aggregation_finder.aggregate_hash
+
+      if definition.need_load_object
+        aggregate_objects = reflection.klass.find(aggregate_hash.values).each_with_object({}) { |o, h| h[o.id] = o }
+        aggregate_hash.keys.each { |key| aggregate_hash[key] = aggregate_objects[aggregate_hash[key]] }
+      end
+
+      @records.each do |record|
+        id = record.send(aggregation_finder.group_by_key)
```

Comment:
`HasManyThroughBelongsTo` would group data by a different key, we should align with that as well.